File: ruby/aoc_6.rb
# frozen_string_literal: true

require "pry"

def part1(data)
  time, distance = data
  h = time.zip(distance).to_h

  h.map { |t, d| t.downto(0).count { |button_t| button_t * (t - button_t) > d } }.reduce(:*)
end

def part2(data)
  time, distance = data.map{ |i| i.join.to_i }
  acc = 0
  time.downto(0) { |button_t| button_t * (time - button_t) > distance ? acc += 1 : next }
  acc
end

data = File.read("input_6.txt").split("\n").map { |l| l.split.grep(/\A\d+\z/).map(&:to_i) }
p part1(data) #=> 512295
p part2(data) #=> 36530883

File: elixir/aoc_6.exs
require IEx

defmodule Day6.Common do
  def read_input(file_path) do
    case File.read(file_path) do
      {:ok, data} ->
        String.split(data, "\n", trim: true)
        |> Enum.map(fn line ->
          line
          |> String.split()
          |> Enum.drop(1)
          |> Enum.map(&String.to_integer/1)
        end)

      {:error, reason} ->
        IO.puts("Error reading file: #{reason}")
        []
    end
  end
end

defmodule Day6.Part1 do
  def solve(data) do
    [time, distance] = data

    time
    |> Enum.zip(distance)
    |> Enum.map(fn {t, d} ->
      0..t
      |> Enum.filter(fn button_t -> button_t * (t - button_t) > d end)
      |> Enum.count()
    end)
    |> Enum.reduce(&*/2)
  end
end

defmodule Day6.Part2 do
  def solve(data) do
    [time, distance] =
      data
      |> Enum.map(fn x ->
        x
        |> Enum.map(&Integer.to_string/1)
        |> Enum.join()
        |> String.to_integer()
      end)

    Enum.reduce(0..time, 0, fn bt, acc ->
      if bt * (time - bt) > distance do
        acc + 1
      else
        acc
      end
    end)
  end
end

data = Day6.Common.read_input("input_6.txt")
# => 512295
IO.puts("Part 1: #{Day6.Part1.solve(data)}")
# # => 36530883
IO.puts("Part 2 : #{Day6.Part2.solve(data)}")
